src/linear_algebra/finsupp.lean

```diff
@@ -313,10 +313,34 @@ finsupp.sum_single_index (f i).map_zero
 
 theorem lsum_symm_apply (f : (α →₀ M) →ₗ[R] N) (x : α) : lsum.symm f x = f.comp (lsingle x) := rfl
 
+section
+variables (M) (R) (X : Type*)
+
+/--
+A slight rearrangement from `lsum` gives us
+the bijection underlying the free-forgetful adjunction for R-modules.
+-/
+noncomputable def hom_equiv : ((X →₀ R) →ₗ[R] M) ≃ (X → M) :=
+(lsum : _ ≃+ _).to_equiv.symm.trans $ equiv.arrow_congr (equiv.refl _)
+{ to_fun := λ f, f 1,
```

Comment:
I think this inner equiv probably deserves a name

Reply:
Indeed. `def ring_lmap_equiv_self : (R →ₗ[R] M) ≃ M`?